cpp/src/copying/pack.cpp

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+
+#include <jit/type.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief The data that is stored as anonymous bytes in the `packed_columns` metadata
+ * field.
+ *
+ * The metadata field of the `packed_columns` struct is simply an array of these.
+ * This struct is exposed here because it is needed by both contiguous_split, pack
+ * and unpack.
+ */
+struct serialized_column {
+  data_type type;
+  size_type size;
+  size_type null_count;
+  int64_t data_offset;       // offset into contiguous data buffer, or -1 if column data is null
+  int64_t null_mask_offset;  // offset into contiguous data buffer, or -1 if column data is null
+  size_type num_children;
+  // padding explicitly so we can force initialize the value to 0, allowing us to do a byte-by-byte
```

Comment:
I'm still not clear on what this is for. Seems a little smelly.

Reply:
The struct ends up padded because it has 64 bit ints and only 3 32 bit ints.  So no matter how you rearrange it, the compiler has to pad out an extra 4 bytes.  Because of this, you can initialize every struct member and still end up with some uninitialized bytes.   It's not the end of the world, but we do end up converting this whole buffer into a stream of bytes (vector<uint8_t>).  

I had added a test that did some explicit comparisons between an incoming buffer and an expected buffer and was surprised to find they didn't match (https://github.com/rapidsai/cudf/blob/0e721e18efe53d2e4020652b0ad4fc93d50c6c3b/cpp/tests/copying/pack_tests.cu#L38), leading me to this.  Since we're treating this as a network-style stream of bytes, it felt cleaner to guarantee bitwise equality, otherwise you could do a series of pack->unpack->pack->unpack->pack->unpack on a the same table and end up with metadata buffers that were physically different, though equivalent.

Reply:
Ah, I see, it's a unique object representation issue. I guess the manual padding is fine as an implementation detail, but I do not think we should make any external guarantees (or encourage callers to rely on this behavior) that packing the same table twice returns bitwise identical metadata. I think that unnecessarily ties our hands.